Add a basic integration of the block email editor to the woo core

triple0t opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/55601

![Screenshot 2025-02-24 at 1 11 02 AM](https://github.com/user-attachments/assets/584fad0a-dbdc-46b0-a75f-b2bda6245a32)


### Email Editor Integration for WooCommerce Core

This PR introduces a basic integration of the block-based email editor into WooCommerce core. This is an experimental feature that provides foundational support for editing transactional emails using the WordPress block editor. **The integration is hidden behind a feature flag and not ready for public consumption**.

### Key Changes

#### Email Editor Integration
- Added basic email editor integration, which will be enabled for WooCommerce transactional emails
- Introduced a new post type `woo_mail` for managing email templates
- Integrated personalization tags specific to WooCommerce (e.g., shopper email, site title)
- Added a basic email template with WooCommerce branding and layout

#### Package Structure
- Added `@woocommerce/email-editor` package dependency (JS-pnpm) and `woocommerce/email-editor` (PHP-composer)
- Moved copy assets commands to the `@woocommerce/plugin-woocommerce` package
- Copy JS build assets to `assets/client/admin/email-editor`  (Optimization of the JS build will be done with https://github.com/woocommerce/woocommerce/issues/55688)

#### New Components
- `Integration.php`: Core integration class managing hooks and post type registration
- `PageRenderer.php`: Handles rendering of the email editor interface
- `PersonalizationTagManager.php`: Manages WooCommerce-specific personalization tags
- `PatternsController.php`: Manages block patterns for email templates
- `TemplatesController.php`: Handles email template registration and management

#### Build System
- Updated build scripts to handle email editor assets
- Added new npm script `email-editor:build:project:copy-assets`

## Notes
- This feature is currently experimental and not ready for public use
- The integration provides basic functionality that will be expanded in future updates
- The email editor UI is accessible through WP Admin but this may change in future iterations

## Related
- Depends on the `@woocommerce/email-editor` package

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the `block_email_editor` experimental flag (in DB, set `woocommerce_feature_block_email_editor_enabled` option to yes).
2. Navigate to WP Admin > Woo Emails
3. Create or edit an email template
4. Verify the block editor loads with WooCommerce template and pattern content

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

triple0t committed:
Add email editor php package to composer.json

triple0t committed:
Add basic email editor integration files

The integration will only be enabled when the block_email_editor feature flag is set

triple0t committed:
Add the pnpm build commands for the email editor js package

I would ideally like to ensure we use the same build system and process as the rest of the monorepo packages but that will increase the scope of this ticket.

Adding webpack config will be done in https://github.com/woocommerce/woocommerce/issues/55688#issue-2866084770

triple0t committed:
Fix errors with the PersonalizationTagManager and typo with the WooMailContentPattern

triple0t committed:
Add JS integration support for the email editor package

triple0t committed:
Add basic integration of the block email editor to WooCommerce core

This commit adds the initial integration of the block email editor to WooCommerce core. The changes include:

- Registering the email editor post type
- Adding personalization tags for WooCommerce emails
- Implementing a basic email template
- Creating patterns for email content
- Setting up service providers and integration hooks

The feature is experimental and not ready for public use. It will only be active when the block_email_editor feature flag is enabled.

triple0t committed:
Merge branch 'trunk' into add-a-basic-integration-of-the-block-email-editor-to-the-woo-core

triple0t committed:
Refactor email editor namespaces and add email patterns and templates

This commit updates the email editor integration by:
- Renaming namespaces for patterns and templates to be more descriptive

The changes improve the organization of the email editor components and provide initial patterns and templates for WooCommerce emails.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/Integration.php:
```diff
@@ -0,0 +1,85 @@
+defined( 'ABSPATH' ) || exit;
+
+class Integration {
+	const EMAIL_POST_TYPE = 'woo_mail';
```
Just a detail. In most cases, we refer to an email as `email` so I would suggest sticking with this and using `woo_email`. In MailPoet, we currently use `mailpoet_email`. I see that when we register the type, we use "Woo Email" as a singular name.

triple0t replied:
Done in https://github.com/woocommerce/woocommerce/pull/55754/commits/c9a010081c8f3a28d53f4bbe61f996abcbfd784b

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/Integration.php:
```diff
@@ -0,0 +1,85 @@
+				],
+				'rewrite' => ['slug' => self::EMAIL_POST_TYPE],
+				'supports' => ['title', 'editor'],
+				'public' => true,
```
Do we want to use `public`? In MailPoet we use public set to `false` in combination with `publicly_queryable` because of previews. If I'm not mistaken with public `true` the email posts will also be searchable in the front end. So we could probably at least set `exclude_from_search` (see [docs](https://developer.wordpress.org/reference/functions/register_post_type/)). Perhaps there might be other effects.

triple0t replied:
I'm keeping this public right now because
1. This is an initial implementation, and it is behind a feature flag; we can update it later
2. Using `public = false` or `exclude_from_search` leads to this error when saving the email. We need to take a closer look at it and fix it or update it later.
![Screenshot 2025-02-25 at 11 32 40 AM](https://github.com/user-attachments/assets/9fa99c10-55b6-4447-b630-6d7957b32044)

costasovo replied:
Thanks for sharing the details. Could you please create an issue for investigating the error so that we don't forget? 

I think the error is related to this code https://github.com/woocommerce/woocommerce/blob/88f1c3c00c9bfc7c51f0b250623c4be761381d52/packages/php/email-editor/src/Engine/Templates/class-templates.php#L140-L167
If I'm not mistaken, we already added the templates registry so we could use it to list email templates instead of [fetching all templates and filtering by the prefix](https://github.com/woocommerce/woocommerce/blob/88f1c3c00c9bfc7c51f0b250623c4be761381d52/packages/php/email-editor/src/Engine/Templates/class-templates.php#L158-L166).

triple0t replied:
I agree with you.
I created a ticket for this here https://github.com/woocommerce/woocommerce/issues/55862

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/Package.php:
```diff
@@ -0,0 +1,63 @@
+	public static function initialize()  {
+		$container = wc_get_container();
+		$container->get( Integration::class );
+	}
```
I don't understand why we need to initialize the Integration class. It doesn't do anything in the constructor, so I struggle to understand why this method is needed.

triple0t replied:
> I don't understand why we need to initialize the Integration class.

We need to initialize the Integration class so we can use the email editor. Otherwise, the editor won't be loaded on the WP admin page.

We're doing it [here](https://github.com/woocommerce/woocommerce/blob/add-a-basic-integration-of-the-block-email-editor-to-the-woo-core/plugins/woocommerce/src/Internal/EmailEditor/Package.php#L37-L47) because we want the integration to load only when the feature flag is enabled

costasovo replied:
Thanks for the explanation. I see that after removing the initialization the integration doesn't trigger. But I, but struggle to understand how that works.
I see in the PHP doc of `$container->get` that it is used to obtain a service. But after this initialization we don't do anything with the service. So there is probably a side-effect I failed to find. Sorry for bothering you with this, but I would like to understand how that works. Could you please explain how that works or point me to some code that would help me to understand?

triple0t replied:
>I see in the PHP doc of $container->get that it is used to obtain a service. But after this initialization we don't do anything with the service. So there is probably a side-effect I failed to find. Sorry for bothering you with this, but I would like to understand how that works. Could you please explain how that works or point me to some code that would help me to understand?

When [we get](https://github.com/woocommerce/woocommerce/blob/b65c371618db855d250385233a05a95896b4daf8/plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php#L86-L126) a service, the init method is [automatically invoked](https://github.com/woocommerce/woocommerce/blob/b65c371618db855d250385233a05a95896b4daf8/plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php#L140-L177), and it's dependent classes (arguments) are [added](https://github.com/woocommerce/woocommerce/blob/b65c371618db855d250385233a05a95896b4daf8/plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php#L154-L170).

So, instead of doing something like
```php
$container->get( Integration::class )->init(...);
```

We can use
```php
$container->get( Integration::class );
```

costasovo replied:
Ha! That's what I failed to notice. It makes perfect sense now. Looks a bit like magic but I understand now. Thank you!

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/PageRenderer.php:
```diff
@@ -0,0 +1,139 @@
+
+		// Email editor rich text JS - Because the Personalization Tags depend on Gutenberg 19.8.0 and higher
+		// the following code replaces used Rich Text for the version containing the necessary changes.
+		$assetsParams = require $emailEditorAssetsPath . 'rich-text.asset.php';
```
Perhaps we could improve the code by adding a check for WP and Gutenberg versions and enqueue the modified RichText script only for WP6.7 without the newer Gutenberg version.

triple0t replied:
I would rather not remove this for the following reasons
1. This is only done for the email editor page
2. It ensures consistent with the other implementation
3. I added a [dependency](https://github.com/woocommerce/woocommerce/pull/55754/commits/c9a010081c8f3a28d53f4bbe61f996abcbfd784b#diff-189848a459676851e7c7e3afca00c5a6b45598df340444cdc2841ed31e9aef98L30) check, and this page (or package) won't be loaded for WP version less than the [recommended](https://github.com/woocommerce/woocommerce/blob/add-a-basic-integration-of-the-block-email-editor-to-the-woo-core/packages/php/email-editor/src/Engine/class-dependency-check.php#L18) version

costasovo replied:
Makes sense. Thank you!

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/patterns/WooMailContentPattern.php:
```diff
@@ -0,0 +1,37 @@
+
+	public function get_title(): string {
+		/* translators: Name of a content pattern used as starting content of an email */
+		return __('Woo Mail Content Pattern', 'woocommerce');
```
We could probably use `Woo Email ...` for consistency.

triple0t replied:
Done in https://github.com/woocommerce/woocommerce/pull/55754/commits/c9a010081c8f3a28d53f4bbe61f996abcbfd784b

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/templates/WooMailTemplate.php:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	public function get_title(): string {
+		return __('WooMailTemplate', 'woocommerce');
```
Shall we rather use the `Woo Email Template"?
I think it would look nicer in UIs like this one:
![Screenshot 2025-02-24 at 13 22 26](https://github.com/user-attachments/assets/0d192b4e-7c89-45ab-8fa4-e2c4fde8649d)

triple0t replied:
Done in https://github.com/woocommerce/woocommerce/pull/55754/commits/c938f12469ee29c0e6b92f752f34600c87ea7648

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/patterns/WooMailContentPattern.php:
```diff
@@ -0,0 +1,37 @@
+use MailPoet\EmailEditor\Engine\Patterns\Abstract_Pattern;
+
+class WooMailContentPattern extends Abstract_Pattern {
+	public $name = 'woo-mail-content-pattern';
```
Sry for nitpicking. I would use WooEmail... and woo-email to be consistent.

triple0t replied:
Done in https://github.com/woocommerce/woocommerce/pull/55754/commits/c9a010081c8f3a28d53f4bbe61f996abcbfd784b

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/templates/WooMailTemplate.php:
```diff
@@ -0,0 +1,36 @@
+class WooMailTemplate {
+
+	public function get_slug(): string {
+		return 'woomailtemplate';
```
Perhaps `wooemailtemplate` ?

triple0t replied:
Done in https://github.com/woocommerce/woocommerce/pull/55754/commits/c9a010081c8f3a28d53f4bbe61f996abcbfd784b

costasovo commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/email-editor-integration/index.ts:
```diff
@@ -0,0 +1,14 @@
+addFilter(
+	'mailpoet_email_editor_send_button_label',
+	NAME_SPACE,
+	() => 'Save WooCommerce email template'
```
I see the string is not wrapped in the translation function, and when testing the editor I noticed it returned to the listing. I assume it is not a temporary button until we figure out how actually handle the block template email activation. Is this assumption correct?

Thinking about it now, I have an idea: we could perhaps use "Activate email" and "Deactivate email," which could set the email as active to replace the associated transactional email. But this is something we will solve in follow-up tickets.

triple0t replied:
> I see the string is not wrapped in the translation function, and when testing the editor I noticed it returned to the listing. I assume it is ~not~ a temporary button until we figure out how actually handle the block template email activation. Is this assumption correct?

Yes, this assumption is correct.
I added a comment clarifying this in https://github.com/woocommerce/woocommerce/pull/55754/commits/c9a010081c8f3a28d53f4bbe61f996abcbfd784b#diff-33bbc2a86fb2e9e21620ad03db75e65a68667caf47d6e4b0c5534877285173d7R13

costasovo reviewed (commented):
@triple0t Thanks for working on the integration. It is cool to see this progress.
I left some comments, mostly details or questions. Sorry for nitpicking.

When I tested the editor, I noticed that it is not possible to save changes in a newly created email.

![Screenshot 2025-02-24 at 13 45 36](https://github.com/user-attachments/assets/c4b93fa8-4f84-4842-87b7-ba43225db897)

Is this a known issue? Shall we create a GH issue for that?

triple0t committed:
Update email editor integration with naming and dependency improvements (use woo_email over woo_mail)

This commit refines the WooCommerce email editor integration by:
- Updating naming conventions for email-related classes and post types
- Adding dependency checks for the email editor
- Adding a temporary comment for the save button label integration

triple0t committed:
Refine save button logic and update WooCommerce email template title

- Adjust save button disabled state to only disable when no edits are present
- Update WooCommerce email template title to use proper capitalization

triple0t commented:
Thank you for the review, @costasovo 

> When I tested the editor, I noticed that it is not possible to save changes in a newly created email.
> Is this a known issue? Shall we create a GH issue for that?

Fixed in https://github.com/woocommerce/woocommerce/pull/55754/commits/c938f12469ee29c0e6b92f752f34600c87ea7648

It's an issue with the editor. Using "Ctrl + S", Waiting for the autosave or using the preview in new tab would have saved the content.

costasovo reviewed (approved):
@triple0t Thanks for adding the integration. Great job! It is an important step for the follow-up tasks.

I followed up on a couple of comments with some questions, but they are not blocking.
Before we merge I have two more additional questions:

1) Is my review enough or shall we ask someone else from Woo for additional review? 
2) The PR doesn't contain any tests. Shouldn't we test the integration? Here are a few things that I think we could test:
- run initialization and check if CPT, patterns, templates are registered
- add a simple E2E test that would open the editor
What do you think about adding tests?

triple0t commented:
Thank you for the review, @costasovo 

> Is my review enough or shall we ask someone else from Woo for additional review?

Yes, this is enough.

> The PR doesn't contain any tests. Shouldn't we test the integration? Here are a few things that I think we could test:
run initialization and check if CPT, patterns, templates are registered
add a simple E2E test that would open the editor
What do you think about adding tests?

I would love to merge this first and then create a ticket to concentrate on the test.

costasovo commented:
> I would love to merge this first and then create a ticket to concentrate on the test.

@triple0t Ok 👍 Let's merge it. Could you please create an issue for adding the tests? 🙏 I assume there is no blocker for adding the tests because we can add tests for the integration directly to the Woo core test suite.

senadir commented:
This code is running too early which I think causes this error 
https://github.com/woocommerce/woocommerce/issues/56113

triple0t commented:
Thank you, @senadir. Fixed in https://github.com/woocommerce/woocommerce/pull/56133
